From 44ad8c5ee6be05c8bfd666af179ad878c12df52c Mon Sep 17 00:00:00 2001
From: aahmed878 <Alsameema_Ahmedansar@comcast.com>
Date: Tue, 14 Mar 2023 05:21:01 +0000
Subject: [PATCH 2/3] qtdemux: Handle protection information in Sample Group
 Description

Override default protection parameters by using defined sample group

https://bugzilla.gnome.org/show_bug.cgi?id=782224
---
 gst/isomp4/qtdemux.c | 79 +++++++++++++++++++++++++++++++++++++++++---
 1 file changed, 75 insertions(+), 4 deletions(-)

diff --git a/gst/isomp4/qtdemux.c b/gst/isomp4/qtdemux.c
index 78ce0aa..6386790 100644
--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -230,7 +230,8 @@ struct _QtDemuxRandomAccessEntry
 struct _QtDemuxCencSampleSetInfo
 {
   GstStructure *default_properties;
-
+  /* overriding default_properties if any */
+  GPtrArray *sample_properties;
   /* @crypto_info holds one GstStructure per sample */
   GPtrArray *crypto_info;
 };
@@ -3720,6 +3721,69 @@ failed:
   }
 }
 
+static void
+qtdemux_prepare_cenc_sample_properties (GstQTDemux * qtdemux,
+    QtDemuxStream * stream)
+{
+  QtDemuxCencSampleSetInfo *info = NULL;
+  QtDemuxSampleToGroup *target_group = NULL;
+  guint i, j;
+  guint32 n_samples = 0;
+
+  g_return_val_if_fail (stream != NULL, NULL);
+  g_return_val_if_fail (stream->protected, NULL);
+  g_return_val_if_fail (stream->protection_scheme_info != NULL, NULL);
+
+  info = (QtDemuxCencSampleSetInfo *) stream->protection_scheme_info;
+
+  if (info->sample_properties)
+    g_ptr_array_free (info->sample_properties, TRUE);
+
+  info->sample_properties = NULL;
+
+  if (!info->default_properties) {
+    GST_ERROR_OBJECT (qtdemux, "Default properties is not defined");
+    return;
+  }
+
+  if (stream->sample_to_group) {
+    GList *iter;
+    for (iter = stream->sample_to_group; iter; iter = g_list_next (iter)) {
+      QtDemuxSampleToGroup *group = (QtDemuxSampleToGroup *) iter->data;
+      if (group->grouping_type == FOURCC_seig) {
+        target_group = group;
+        break;
+      }
+    }
+  }
+
+  if (!target_group)
+    return;
+
+  for (i = 0; i < target_group->sbgp_entries_length; i++) {
+    QtDemuxSampleToGroupEntry *group_entry = &target_group->sbgp_entries[i];
+    n_samples += group_entry->sample_count;
+    if (!info->sample_properties) {
+      info->sample_properties =
+          g_ptr_array_new_full (n_samples,
+          (GDestroyNotify) qtdemux_gst_structure_free);
+    } else {
+      g_ptr_array_set_size (info->sample_properties, n_samples);
+    }
+
+    for (j = 0; j < group_entry->sample_count; j++) {
+      GstStructure *properties;
+
+      /* group->data might be NULL if group_description_index was zero */
+      properties = group_entry->data ?
+          group_entry->data : info->default_properties;
+      g_ptr_array_add (info->sample_properties,
+          gst_structure_copy (properties));
+    }
+  }
+}
+
+
 /* Returns a pointer to a GstStructure containing the properties of
  * the stream sample identified by @sample_index. The caller must unref
  * the returned object after use. Returns NULL if unsuccessful. */
@@ -3728,6 +3792,7 @@ qtdemux_get_cenc_sample_properties (GstQTDemux * qtdemux,
     QtDemuxStream * stream, guint sample_index)
 {
   QtDemuxCencSampleSetInfo *info = NULL;
+  GstStructure *properties = NULL;
 
   g_return_val_if_fail (stream != NULL, NULL);
   g_return_val_if_fail (stream->protected, NULL);
@@ -3735,9 +3800,13 @@ qtdemux_get_cenc_sample_properties (GstQTDemux * qtdemux,
 
   info = (QtDemuxCencSampleSetInfo *) stream->protection_scheme_info;
 
-  /* Currently, cenc properties for groups of samples are not supported, so
-   * simply return a copy of the default sample properties */
-  return gst_structure_copy (info->default_properties);
+  if (info->sample_properties && sample_index < info->sample_properties->len) {
+    properties = g_ptr_array_index (info->sample_properties, sample_index);
+    g_ptr_array_index (info->sample_properties, sample_index) = NULL;
+  }
+  if (!properties)
+    properties = gst_structure_copy (info->default_properties);
+  return properties;
 }
 
 /* Parses the sizes of sample auxiliary information contained within a stream,
@@ -3919,6 +3988,8 @@ qtdemux_parse_cenc_aux_info (GstQTDemux * qtdemux, QtDemuxStream * stream,
     /* Everything now belongs to the new array */
     g_ptr_array_free (old_crypto_info, TRUE);
   }
+  /* FIXME: how to handle old_crypto_info ? */
+  qtdemux_prepare_cenc_sample_properties (qtdemux, stream);
 
   for (i = 0; i < sample_count; ++i) {
     GstStructure *properties;
-- 
2.24.3

